### ISSUE TITLE
Version requirements for Python dependencies too strict

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#44654]
**System information**
- TensorFlow version (you are using): 2.4.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Up until 2.3.1 the version requirements in the setup.py file where of the form `foo >=x.y` or `foo >=x.y,<z.q`, i.e. there was a minimum and a potential maximum version specified. This allows for TensorFlow to be installed alongside other Python packages with own requirements.
However https://github.com/tensorflow/tensorflow/commit/aafe25d3f97d645bca91b34a9476eb770d1abf90 changed that significantly and IMO to the worse: It basically pins a specific major and minor version.

I think this is overly eager. For example TF works fine with protobuf 3.9.2 or 3.10 but now requires 3.13.x. Similar the wheel version doesn't really matter much (a 0.26 is already enough) but it now requires a 0.35.x

This is really bad for forward and backward compatibility. For example the system Python (e.g. on a HPC cluster, which can't be changed) could include a newer wheel (say 0.36.0) for e.g. security reasons. Now TF cannot be installed in that environment any longer even though there is no actual reason but that arbitrary limitation to a version which might have been the most recent at some point in time. Also an already installed wheel 0.33.0 can't be used although everything would work.

And also cross compatibility is affected. Let's say a user needs to install a software which needs a feature from wheel 0.36, this can now not be installed because TF limited the version to 0.35

**Will this change the current api? How?**

No.

**Who will benefit with this feature?**

Everyone using TF in existing environments and/or with other packages

**Any Other info.**

My suggestion is to revert the version list back to what it was using min and max versions and be quite liberal there too (e.g. numpy >=1.16 works just fine and even just limiting to either pre 1.19 or post 1.19 would exclude whole classes of other software). The `~=` notation can be used where it makes sense, e.g. for `tensorboard` whose major and minor version must match the TF version.

**Comment by mihaimaruseac** [comment]
Based on my reading, `wheel ~= 0.35` means any `wheel` version from 0.35, 0.36, ... up to 1.0 (exclusive) can be installed, as the resolver picks the latest version from cache that still satisfies all the constraints.

`~x.y` means every version after `x.y` but before `x+1.0`.

We should not be using `>= x.y` because that harms releasing patch releases. We've had issues when a dependency released a new version and suddenly tests started failing.

**Comment by Flamefire** [comment]
This is true, wheel is currently set at >=0.36 < 1.0. However that is needlessly to high. E.g. on our cluster we have wheel 0.33.6 installed (such stuff usually gets installed once and updated once a year when a new toolchain generation gets installed which changes pretty much everything) and that version would well be enough. There are no features used which are not present in 0.33 so the requirement is too strict and was likely simply generated by simply using whatever was the latest version at that time. IMO the better approach would be to translate the old version requirements (in this case: `'wheel >= 0.26'`) to the new feature which would be `wheel ~= 0.26`.
For `six` it is even worse: Before: `'six >= 1.12.0'`, Now: `six~=1.15.0`, so this is restricted to the 1.15.x versions and 1.12 or 1.16 or not accepted which I see no reasoning for.

So my point is not that it is wrong to restrict the major version, but that indistinctly restricting the minor version and raising the minimum version without a reason should be avoided to not break peoples environments.
To me it looks like an oversight in a search&replace action not an intentional change.

**Comment by mihaimaruseac** [comment]
That makes sense. I will try expanding the bounds.

The change was to go ahead of the new pip resolver, to not cause errors. That's why we picked restrictive bounds for now and will slowly expand, based on this feedback.

**Comment by fridex** [comment]
> That makes sense. I will try expanding the bounds.
> 
> The change was to go ahead of the new pip resolver, to not cause errors. That's why we picked restrictive bounds for now and will slowly expand, based on this feedback.

Do you have any specific testsuite to run tensorflow with changes in requirements? It might be a computationally intensive task to check all the TF parts so having a testsuite specific for this case might be nice. We have [a component that can generate software stacks following programmed rules](https://github.com/thoth-station/adviser) (a programable sw stack resolver). We spotted more issues in TF dependencies - see for example https://developers.redhat.com/blog/2020/09/30/ai-software-stack-inspection-with-thoth-and-tensorflow/

**Comment by mihaimaruseac** [comment]
There is no simpler test-suite. We are building the pip and testing that it can be installed with the newer pip resolver.

Also, linking #44467 to this one since it is mostly caused by overly-large bounds.

**Comment by mihaimaruseac** [comment]
It seems we will need to increase upper bounds of dependencies to support python3.9 without overcomplicating `setup.py`

**Comment by mihaimaruseac** [comment]
There is also https://github.com/tensorflow/tensorflow/issues/35027#issuecomment-734791293 that suggests another issue with too lax dependencies.

**Comment by Flamefire** [comment]
> There is also #35027 (comment) that suggests another issue with too lax dependencies.

Not really. The users reported an incompatibility with absl-py < 0.9 but TF has required >= 0.9 before the change mentioned here, so that isn't a recent issue at least

**Comment by mihaimaruseac** [comment]
Agree, it's just a datapoint to not go below for absl.

**Comment by mihaimaruseac** [comment]
Another datapoint (for when we can finally tackle this): numpy should be restricted to be either above 1.20 (and drop py36 support) or below 1.19.

Breaking change: https://numpy.org/doc/stable/release/1.20.0-notes.html#the-pyarray-descrcheck-macro-is-modified

**Comment by Flamefire** [comment]
> Breaking change: https://numpy.org/doc/stable/release/1.20.0-notes.html#the-pyarray-descrcheck-macro-is-modified

FWIW: This macro is not used in TF and a portable solution exists (use the explicit function)

**Comment by mihaimaruseac** [comment]
Most builds were failing yesterday because of numpy 1.20. We are currently limiting all dependencies/numpy so that we don't get numpy 1.20 in the CI.

**Comment by edrozenberg** [comment]
> Most builds were failing yesterday because of numpy 1.20. We are currently limiting all dependencies/numpy so that we don't get numpy 1.20 in the CI.

Also ran into 1.20 not being supported. Compiled tf 2.4.1 and the tf wheel wants to remove my numpy 1.20 and replace it with 1.19.5. In addition I have to rebuild at least `pandas` - my pandas was built w/ numpy 1.20 installed, and trying to import pandas now says `ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 80 from PyObject`.

**Comment by fridex** [comment]
> Another datapoint (for when we can finally tackle this): numpy should be restricted to be either above 1.20 (and drop py36 support) or below 1.19.

@mihaimaruseac If I read this correctly, should TensorFlow=2.4.x use numpy>1.20 for Python>=3.7 and numpy<1.19 for Python==3.6?

**Comment by mihaimaruseac** [comment]
@fridex sure, but this will have diverging behavior (also, 1.19.5 is good)
@edrozenberg that's why we're limiting numpy to ~1.19.2 to get all the patches on 1.19 series.

Once we can drop support for python 3.6 (or numpy adds it back), we'll invest into moving to `> 1.20`)

**Comment by fridex** [comment]
> @fridex sure, but this will have diverging behavior (also, 1.19.5 is good)

@mihaimaruseac , thanks for the clarification. I'm not sure if https://github.com/tensorflow/tensorflow/issues/47117 is somehow associated as there were issues with numpy==1.19.5. Is it OK to always run numpy==1.19.5 with TensorFlow 2.4 or is there a specific issue with numpy=1.19.5? Sorry, it's not directly clear to me from the issue and logs, especially https://github.com/tensorflow/tensorflow/issues/47117 confuses me. Thanks for your time.

**Comment by mihaimaruseac** [comment]
1.19.5 is good. We have no successful 1.20 build (we did not try to get one)

**Comment by dciborow** [comment]
>  numpy >=1.16 works just fine and even just limiting to either pre 1.19 or post 1.19 would exclude whole classes of other software). T

This is exactly our issue.

**Comment by Dobiasd** [comment]
Just ran into another problem caused by TensorFlow using `~=` instead of `>=` for its dependencies:

```
ERROR: Cannot install -r requirements.txt (line 11), -r requirements.txt (line 17) and -r requirements.txt (line 33) because these package versions have conflicting dependencies.
The conflict is caused by:
    mypy 0.910 depends on typing-extensions>=3.7.4
    pylint 2.11.1 depends on typing-extensions>=3.10.0; python_version < "3.10"
    tensorflow 2.6.0 depends on typing-extensions~=3.7.4
```

So basically, we can't update `pylint` because of that. :neutral_face:

I think it would be *much* better if TensorFlow would require `typing-extensions>=3.7.4` instead of `typing-extensions~=3.7.4`.

**Comment by mihaimaruseac** [comment]
This should be fixed now, I think?

**Comment by Dobiasd** [comment]
> This should be fixed now, I think?

Just tried this minimal `Dockerfile` with the latest versions of `pylint` and `tensorflow`:

```Dockerfile
FROM python:3.9
RUN pip install pylint==2.11.1 tensorflow==2.6.0
```

Output:

```
ERROR: Cannot install pylint==2.11.1 and tensorflow==2.6.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    pylint 2.11.1 depends on typing-extensions>=3.10.0; python_version < "3.10"
    tensorflow 2.6.0 depends on typing-extensions~=3.7.4
```

So, no, it's not fixed.

**Comment by mihaimaruseac** [comment]
Oh, it's fixed in nightly, not in the released TF.

**Comment by Dobiasd** [comment]
Oh, sorry. And thanks! :slightly_smiling_face:

The following works nicely:

```Dockerfile
FROM python:3.9
RUN pip install pylint==2.11.1 tf-nightly==2.8.0.dev20211005
```

:tada:

**Comment by mihaimaruseac** [comment]
We'll cherry-pick this on the 2.7 branch and from RC1 (RC0 is too late as the builds started) this should be the default

**Comment by mihaimaruseac** [comment]
This should be fully fixed now.

**Comment by shlomiken** [comment]
why this issue is closed ? even working with 2.6 version still requires typing extensions ~3.7.4 which many libraries cannot work with.  why all other libraries can be more loose about the version ?
```
There are incompatible versions in the resolved dependencies:
  typing-extensions~=3.7.4 (from tensorflow==2.6.0->-r requirements.in (line 8))
  typing-extensions>=4.1.0 (from mypy-boto3-redshift-data==1.21.34->apache-airflow-providers-amazon==1!3.2.0->-r requirements.in (line 21))
  typing_extensions (from cattrs==1.10.0->apache-airflow==1!2.2.5+astro.2->-r requirements.in (line 30))
  typing-extensions>=3.7.4.3 (from gitpython==3.1.27->mlflow==1.24.0->-r requirements.in (line 4))
  typing-extensions>=3.6.4 (from importlib-metadata==4.11.3->moto==3.1.4->-r requirements.in (line 52))
  typing-extensions>=3.7.4 (from apache-airflow==1!2.2.5+astro.2->-r requirements.in (line 30))
  typing-extensions<5.0,>=4.0.0 (from rich==12.2.0->apache-airflow==1!2.2.5+astro.2->-r requirements.in (line 30))
  typing-extensions>=4.1.0 (from mypy-boto3-rds==1.21.34->apache-airflow-providers-amazon==1!3.2.0->-r requirements.in (line 21))
  typing-extensions (from kiwisolver==1.4.2->matplotlib==3.5.1->-r requirements.in (line 19))

```

**Comment by mihaimaruseac** [comment]
Because this is fixed form TF 2.7 onwards

**Comment by shlomiken** [comment]
> th. why all other libraries can be more loose about th

:+1:

**Comment by shlomiken** [comment]
> Because this is fixed form TF 2.7 onwards

:+1:
